console: Add Azure & Keycloak docs (#1429)
redpanda-data/documentation: modules/manage/pages/cluster-maintenance/disk-utilization.adoc

[[set-time-based-retention]]
=== Set time-based retention

Redpanda enforces time-based retention to manage log segment deletion according to record timestamps. Messages are eligible for deletion when their age exceeds the value specified in `log_retention_ms` (the cluster-level property) or `retention.ms` (the topic-level property). Only closed segments are eligible for deletion and all messages in a closed segment must exceed the age limit before Redpanda considers the segment for cleanup. If `retention.ms` is not set at the topic level, the topic inherits the `log_retention_ms` setting.

> Starting in v25.3, new clusters will use the maximum record timestamp (`max_timestamp`, the default) from each record batch to determine when to delete segments. By validating and using `max_timestamp` directly from record batches, Redpanda ensures:

Reviewer: "By validating" is vague- timestamps are validated (i.e records are either accepted or rejected) by using the properties `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms` to ensure records are not produced with timestamps that are too far in the future or the past (thereby having detrimental/unexpected effects on retention enforcement) in `v25.3` and beyond.

> Starting in v25.3, new clusters will use the maximum record timestamp (`max_timestamp`, the default) from each record batch to determine when to delete segments. By validating and using `max_timestamp` directly from record batches, Redpanda ensures:

Reviewer: Also may not want to prefix this paragraph with "Starting in v25.3", since we already prefix the next paragraph with "created prior to v25.3". Up to you!

Reviewer 2: "Starting in v25.3" is a convention we use to ensure users know what release a feature first appears in. 

The bit "create prior to v25.3" seems necessary because we are telling users how to convert those clusters to use/benefit from this new behavior. At least that is how I understand it. Am I off the mark here?

Reviewer: ack, fine as is then 🙂

> Starting in v25.3, new clusters will use the maximum record timestamp (`max_timestamp`, the default) from each record batch to determine when to delete segments. By validating and using `max_timestamp` directly from record batches, Redpanda ensures:

Reviewer 2: ```suggestion
Starting in v25.3, Redpanda validates timestamps using the properties `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms`. These settings ensure that records are only accepted if their timestamps are within an allowed range, and prevent retention issues caused by records with timestamps that are too far in the past or future. By validating directly from record batches, Redpanda ensures:
```

* Accurate, timestamp-based retention across clusters
* Predictable and consistent data deletion behavior
* Proper retention alignment for replicated and disaster recovery use cases

> Clusters created prior to v25.3 may continue to use the broker timestamp (`broker_timestamp`), which is populated when the broker first receives the record. However, you can manually opt older clusters in to this new `max_timestamp` behavior using the Admin API:

Reviewer 2: ```suggestion
Clusters created prior to v25.3 may continue to use the broker timestamp (`broker_timestamp`), which is populated when the broker first receives the record. 

You can also manually opt in your pre-v25.3 clusters to this new `max_timestamp` behavior using the Admin API:
```

[source,bash]
----
curl  --request PUT http://localhost:9644/v1/features/validated_batch_timestamps -d '{"state":"active"}'
----

> IMPORTANT: When manually opting in older clusters, it is particularly important to specify the destination cluster for users of MirrorMaker2 and Shadowing, as without this feature flag active, older clusters created by MirrorMaker2 will have retention enforced from the date of cluster creation rather than the date of the data. You can see this when you compare the data present on the original cluster versus the newly-replicated cluster, as it will show a difference in retention enforcement.

Reviewer: > IMPORTANT: When manually opting in older clusters, it is particularly important to specify the destination cluster for users of MirrorMaker2 and Shadowing

This sentence is a bit unclear- we want to say that it is particularly important for users of MM2 and Shadowing to enable this feature flag on their destination clusters.

> as without this feature flag active, older clusters created by MirrorMaker2

"older clusters" once again needs to indicate pre-`v25.3` explicitly to not leave any ambiguity on the table I think.

> IMPORTANT: When manually opting in older clusters, it is particularly important to specify the destination cluster for users of MirrorMaker2 and Shadowing, as without this feature flag active, older clusters created by MirrorMaker2 will have retention enforced from the date of cluster creation rather than the date of the data. You can see this when you compare the data present on the original cluster versus the newly-replicated cluster, as it will show a difference in retention enforcement.

Reviewer 2: ```suggestion
IMPORTANT: When manually opting in pre-v25.3 clusters, it is particularly important for users of MirrorMaker2 and Shadowing to enable this feature flag on their destination clusters. Without this feature flag, active pre-v25.3 clusters created by MirrorMaker2 will have retention enforced from the date of cluster creation rather than the date of the data. You can see this if you compare the data present on the original cluster versus the newly-replicated cluster, as it will show a difference in retention enforcement.
```

NOTE: Make sure that all of the produced data sent to `redpanda` pre-validation checks are correct with respect to having properly set `max_timestamp` in their record batches. Opting pre-v25.3 clusters with bad timestamps into this behavior may result in:
* Data reclaimed earlier than you expect when using `broker_timestamp`-based retention due to unset or low `max_timestamp` values

> * Retention could be blocked due to high `max_timestamp` at some point in the future.

Reviewer 2: ```suggestion
NOTE: Make sure that all of the produced data sent to `redpanda` pre-validation checks is correct with respect to having properly set `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms` in each record batch. If these timestamps are missing or incorrect:
* Data reclaimed earlier than you expect when using `broker_timestamp`-based retention due to unset or low `log_message_timestamp_before_max_ms` values
* Retention could be blocked due to high `log_message_timestamp_after_max_ms` at some point in the future.
```

You can also manually opt in your pre-v25.3 clusters to this new `max_timestamp` behavior using the Admin API:

[source,bash]
----
curl  --request PUT http://localhost:9644/v1/features/validated_batch_timestamps -d '{"state":"active"}'
----

IMPORTANT: When manually opting in pre-v25.3 clusters, it is particularly important for users of MirrorMaker2 and Shadowing to enable this feature flag on their destination clusters. Without this feature flag, active pre-v25.3 clusters created by MirrorMaker2 will have retention enforced from the date of cluster creation rather than the date of the data. You can see this if you compare the data present on the original cluster versus the newly-replicated cluster, as it will show a difference in retention enforcement.

NOTE: Make sure that all of the produced data sent to `redpanda` pre-validation checks is correct with respect to having properly set `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms` in each record batch. If these timestamps are missing or incorrect:

> * Data reclaimed earlier than you expect when using `broker_timestamp`-based retention due to unset or low `log_message_timestamp_before_max_ms` values

Reviewer 2: ```suggestion
* Data reclaimed earlier than you expect when using `broker_timestamp`-based retention due to unset or low `max_timestamp` values
```

> * Retention could be blocked due to high `max_timestamp` at some point in the future.

Reviewer 2: ```suggestion
[NOTE]
====
Make sure that all of the produced data sent to `redpanda` pre-validation checks is correct with respect to having properly set `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms` in each record batch. If these timestamps are missing or incorrect:
- Data reclaimed earlier than you expect when using `broker_timestamp`-based retention due to unset or low `max_timestamp` values
- Retention could be blocked due to high `max_timestamp` at some point in the future.
====
```

> * Retention could be blocked due to high `log_message_timestamp_after_max_ms` at some point in the future.

Reviewer 2: ```suggestion
* Retention could be blocked due to high `max_timestamp` at some point in the future.
```

> * Retention could be blocked due to high `log_message_timestamp_after_max_ms` at some point in the future.

Reviewer: low/high `max_timestamp` was correct here

* Accurate, timestamp-based retention across clusters
* Predictable and consistent data deletion behavior
* Proper retention alignment for replicated and disaster recovery use cases

Clusters created prior to v25.3 may continue to use the broker timestamp (`broker_timestamp`), which is populated when the broker first receives the record. 

You can also manually opt in your pre-v25.3 clusters to this new `max_timestamp` behavior using the Admin API:

[source,bash]
----
curl  --request PUT http://localhost:9644/v1/features/validated_batch_timestamps -d '{"state":"active"}'
----

IMPORTANT: When manually opting in pre-v25.3 clusters, it is particularly important for users of MirrorMaker2 and Shadowing to enable this feature flag on their destination clusters. Without this feature flag, active pre-v25.3 clusters created by MirrorMaker2 will have retention enforced from the date of cluster creation rather than the date of the data. You can see this if you compare the data present on the original cluster versus the newly-replicated cluster, as it will show a difference in retention enforcement.

[NOTE]

> ====
> Make sure that all of the produced data sent to `redpanda` pre-validation checks is correct with respect to having properly set `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms` in each record batch. If these timestamps are missing or incorrect:
> - Data reclaimed earlier than you expect when using `broker_timestamp`-based retention due to unset or low `max_timestamp` values
> - Retention could be blocked due to high `max_timestamp` at some point in the future.

Reviewer 2: ```suggestion
Make sure that all of the produced data sent to `redpanda` pre-validation checks is correct with respect to having properly set `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms` in each record batch. If these timestamps are missing or incorrect:

- Data reclaimed earlier than you expect when using `broker_timestamp`-based retention due to unset or low `max_timestamp` values
- Retention could be blocked due to high `max_timestamp` at some point in the future.
```

[[set-time-based-retention]]
=== Set time-based retention

> Redpanda enforces time-based retention to manage log segment deletion according to record timestamps. Messages are eligible for deletion when their age exceeds the value specified in `log_retention_ms` (the cluster-level property) or `retention.ms` (the topic-level property). Only closed segments are eligible for deletion and all messages in a closed segment must exceed the age limit before Redpanda considers the segment for cleanup. If `retention.ms` is not set at the topic level, the topic inherits the `log_retention_ms` setting.

Reviewer 3: Consider one of the two: 

```suggestion
Redpanda deletes log segments based on record timestamps. Messages are eligible for deletion when their age exceeds the value specified in `log_retention_ms` (the cluster-level property) or `retention.ms` (the topic-level property). Only closed segments are eligible for deletion and all messages in a closed segment must exceed the age limit before Redpanda considers the segment for cleanup. If `retention.ms` is not set at the topic level, the topic inherits the `log_retention_ms` setting.
```

or

```suggestion
Redpanda deletes log segments based on message age. Messages are eligible for deletion when their age exceeds the value specified in `log_retention_ms` (the cluster-level property) or `retention.ms` (the topic-level property). Only closed segments are eligible for deletion and all messages in a closed segment must exceed the age limit before Redpanda considers the segment for cleanup. If `retention.ms` is not set at the topic level, the topic inherits the `log_retention_ms` setting.
```

Reviewer 2: I think that while it's true that timestamps are inherently time-based, that statement clarifies _how_ time-based retention works in RP: retention is enforced specifically by evaluating the timestamps in records, not just by segment age or other time metrics. I think that this distinction is important for users who may be familiar with other systems that use different mechanisms for time-based retention.

Starting in v25.3, Redpanda validates timestamps using the properties `log_message_timestamp_before_max_ms` and `log_message_timestamp_after_max_ms`. These settings ensure that records are only accepted if their timestamps are within an allowed range, and prevent retention issues caused by records with timestamps that are too far in the past or future. By validating directly from record batches, Redpanda ensures:

* Accurate, timestamp-based retention across clusters
* Predictable and consistent data deletion behavior
* Proper retention alignment for replicated and disaster recovery use cases

> Clusters created prior to v25.3 may continue to use the broker timestamp (`broker_timestamp`), which is populated when the broker first receives the record. 

Reviewer 3: ```suggestion
Clusters running versions prior to v25.3 may continue to use the broker timestamp (`broker_timestamp`), which is populated when the broker first receives the record. 
```

creation time doens`t always imply following a version. It's a metter of picking the version rather than when you created the cluster.

> You can also manually opt in your pre-v25.3 clusters to this new `max_timestamp` behavior using the Admin API:

Reviewer 3: ```suggestion
You can also manually opt in clusters running versions prior to v25.3  to this new `max_timestamp` behavior using the Admin API:
```

[source,bash]
----
curl  --request PUT http://localhost:9644/v1/features/validated_batch_timestamps -d '{"state":"active"}'
----

> IMPORTANT: When manually opting in pre-v25.3 clusters, it is particularly important for users of MirrorMaker2 and Shadowing to enable this feature flag on their destination clusters. Without this feature flag, active pre-v25.3 clusters created by MirrorMaker2 will have retention enforced from the date of cluster creation rather than the date of the data. You can see this if you compare the data present on the original cluster versus the newly-replicated cluster, as it will show a difference in retention enforcement.

Reviewer 2: ```suggestion
IMPORTANT: When manually opting in clusters running versions prior to v25.3, it is particularly important for users of MirrorMaker2 and Shadowing to enable this feature flag on their destination clusters. Without this feature flag, active clusters created by MirrorMaker2 that are running versions prior to v25.3 will have retention enforced from the date of cluster creation rather than the date of the data. You can see this if you compare the data present on the original cluster versus the newly-replicated cluster, as it will show a difference in retention enforcement.
```
